File: LeetCode/p56.py
from operator import index
from typing import List


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if len(intervals) <= 1:
            return intervals

        intervals.sort(key = lambda x: x[0])
        
        res = [intervals[0]]

        for start, end in intervals[1:]:
            last = res[-1][1] 
            if start <= last:
                res[-1][1] = max(last, end)
            else:
                res.append[start, end]


        return res



print("Solution", Solution().merge([[1,4],[2,3]]))

File: LeetCode-Java/p56.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public int[][] merge(int[][] intervals) {
        if (intervals.length <= 1) return intervals;
        
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        
        List<int[]> result = new ArrayList<>();
        int[] prev = intervals[0];
        
        for (int i = 1; i < intervals.length; i++) {
            int[] curr = intervals[i];
            
            if (curr[0] <= prev[1]) {
                prev[1] = Math.max(prev[1], curr[1]);
            } else {
                result.add(prev);
                prev = curr;
            }
        }
        
        result.add(prev);
        
        return result.toArray(new int[result.size()][]);
    }
}
